doc/src/devdocs/compiler.md

```diff
@@ -0,0 +1,119 @@
+# High-level Overview of the Native-Code Generation Process
+
+
+    <placeholder>
+
+
+## Representation of Pointers
+
+When emitting code to an object file, pointers will be emitted as relocations.
+The deserialization code will ensure any object that pointed to one of these constants
+gets recreated and contains the right runtime pointer.
+
+Otherwise, they will be emitted as literal constants.
+
+To emit one of these objects, call `literal_pointer_val`.
+It'll handle tracking the Julia value and the LLVM global,
+ensuring they are valid both for the current runtime and after deserialization.
+
+When emitted into the object file, these globals are stored as references
+in a large `gvals` table. This allows the deserializer to reference them by index,
+and implement a custom manual GOT-like mechanism to restore them.
+
+Function pointers are handled similarly.
+They are stored as values in a large `fvals` table.
+Like globals, this allows the deserializer to reference them by index.
+
+Note that extern functions are handled separately,
+with names, via the usual symbol resolution mechanism in the linker.
+
+Note too that ccall functions are also handled separately,
+via a manual GOT + PLT.
+
+
+## Representation of Intermediate Values
+
+Values are passed around in a `jl_cgval_t` struct.
+This represents an R-value, and includes enough information to
+determine how to assign or pass it somewhere.
+
+They are created via one of the helper constructors, usually:
+`mark_julia_type` (for immediate values) and `mark_julia_slot` (for pointers to values).
+
+The function `convert_julia_type` can transform between any two types.
+When it returns and returns an `cgval.typ` set to `typ`.
+It'll cast the object to the requested representation.
+It'll make boxes, allocate stack copies, and compute tagged unions as
+needed to perform the request.
+
+By contrast `update_julia_type` will change `cgval.typ` to `typ`,
+only if it can be done at zero-cost (i.e. without emitting any code).
+
+
+## Union representation
+
+Inferred union types may be stack allocated via a tagged type representation.
+
+The primitive routines that need to be able to handle tagged unions are:
+- remark-type
+- load-local
+- store-local
+- isa
+- is
+- emit_typeof
+- emit_sizeof
+- boxed
+- unbox
+- specialized cc-ret
```

Comment:
I'm generally reluctant to code-format every word that also appears in code. In part because it is not really standard grammar, and in part because I try to avoid having every other word be bracketed :P